### PR TITLE
Feature: Prompt to update

### PR DESCRIPTION
**Pull request by SandPod**
Continuation of this PR: https://github.com/serverpod/serverpod/pull/619.
Issue: https://github.com/serverpod/serverpod/issues/611

Added feature that will show prompt to update message when a new version of Serverpod CLI is

Screenshot 2023-07-06 at 12 22 07
 available.

## This Feature
- Informs CLI users when a new version of Serverpod is available.
- Fetches the information about the latest version from Pub Dev.
- Stores the latest version with a validity timestamp of 1 day in the shared local Serverpod storage folder.
- If fetching the latest information from Pub Dev fails, doesn't try a refetch within the next hour.

## Additional changes
- Adds two libraries, mockito and build_runner as dev dependencies to the Serverpod_cli. These were required to enabling testing for the calls to pub.dev.
- Locks build_runner to version 2.3.3 since that is the last version that supports dart 2.19.0 which is our lowest supported version.
- Re-builds the freezed_custom_class in the serverpod_test_shared folder with the lower version of build_runner.

## Guidance for review
Best to review the end state of this PR since some features were added and later on remote. The later git history is also a bit scrambled because of some trial and error with the git workflow runner.

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

If you need help, consider asking for advice on the [discussion board](https://github.com/serverpod/serverpod/discussions).

### CODE REVIEW

File: tools/serverpod_cli/bin/serverpod_cli.dart [hunk 1]
```diff
@@ -204,9 +206,10 @@ Future<void> _main(List<String> args) async {
   parser.addCommand(cmdAnalyzePubspecs, analyzePubspecs);
 
   ArgResults results;
+  var devPrint = true;
   try {
     results = parser.parse(args);
-    bool devPrint = results['development-print'];
+    devPrint = results['development-print'];
```
Comment (SandPod) [inline tools/serverpod_cli/bin/serverpod_cli.dart:212]: Chore: This has more values than just development-prints now. We should refactor the meaning of the variable so that it silences all prints. Will do that as part of solving this issue: https://github.com/serverpod/serverpod/issues/1042

File: tools/serverpod_cli/lib/src/update_prompt/prompt_to_update.dart [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  }
+}
+
+void _printPrompt(Version version) {
+  print('');
+  print(_promptMessage(version));
+  print('');
+}
+
+String _promptMessage(Version version) {
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/update_prompt/prompt_to_update.dart:20]: Using the version of the print message that @dmytro-virych created. We will revisit the output style once we have a centralised log service.

File: tools/serverpod_cli/lib/src/util/latest_cli_version.dart [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+      );
+    }
+
+    await resourceManager.storeLatestCliVersion(versionArtefact,
+        localStoragePath: optionalLocalStoragePath);
+    return returnVersion;
+  }
+
+  Future<Version?> _tryFetchLatestCliVersion() async {
+    try {
```
Comment (vlidholt) [inline tools/serverpod_cli/lib/src/util/latest_cli_version.dart:86]: Have you checked if there is an official dart package for doing this? I feel like there may be some package for interacting with pub.dev. (Could be safer to use Google's own stuff)

**Review by vlidholt — CHANGES_REQUESTED** [review]
My biggest concern is the introduction of build_runner. I think we probably should have a discussion around that on Monday. If we do add it, there need to also be tests that check that build_runner has been run and all generated classes are up-to-date. It may include more changes too. Also, referring to a specific version that is outdated seem to not be a good idea.

The other parts looks good with some minor changes. :)

File: templates/pubspecs/tools/serverpod_cli/pubspec.yaml [hunk 4]
```diff
@@ -42,7 +42,9 @@ dependencies:
   pub_semver: ^2.1.4
 
 dev_dependencies:
+  build_runner: 2.3.3
```
[the remark this answers is not in the document]
Comment (SandPod) [reply templates/pubspecs/tools/serverpod_cli/pubspec.yaml:45]: Let's discuss it when you are back.

Right now it is only used to create a mock class for tests. The generated file is also checked in so there shouldn't be a need to re-run the file generation for the test to work. We could also implement the mock class ourselves to remove this dependency.

The strict version is needed because we support Dart SDK 2.19 and build_runner 2.3.3 was the last version that supported Dart SDK 2.14 and above. The newer versions only support Dart SDK 3.0. Source: https://pub.dev/packages/build_runner/versions

File: tools/serverpod_cli/lib/src/util/latest_cli_version.dart (hunk 3 above)
Comment (SandPod) [reply tools/serverpod_cli/lib/src/util/latest_cli_version.dart:86]: As far as I could tell, the dart team unfortunately doesn't have an official tool for this. I thought [cli_util](https://pub.dev/packages/cli_util) would help but it doesn't support fetching the version.

These are other packages I found:
- [pub_updater](https://pub.dev/packages/pub_updater)
- [cli_notify](https://pub.dev/packages/cli_notify)
- [update](https://pub.dev/packages/update)
They have unfortunately not reached version 1 yet so I'm not sure if it is worth adding a dependency for such simple logic in that case.

As a reference the flutter team has implemented their [own version of this](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/version.dart), but then checking towards a GitHub repo instead of pub.dev:

File: tools/serverpod_cli/lib/src/downloads/resource_manager.dart [hunk 5]
```diff
@@ -40,22 +43,85 @@ class ResourceManager {
...
+  }) async {
+    localStoragePath ??= localStorageDirectory.path;
+    var latestCliVersionFile = File(p.join(
+      localStoragePath,
+      ResourceManagerConstants.latestVersionFilePath,
+    ));
+
+    try {
+      if (latestCliVersionFile.existsSync()) {
+        var yaml = loadYaml(
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/lib/src/downloads/resource_manager.dart:97]: Sounds good! 👍

File: tools/serverpod_cli/lib/src/downloads/resource_manager.dart [hunk 6]
```diff
@@ -91,3 +157,10 @@ class ResourceManager {
     print('Download complete.\n');
   }
 }
+
+class LatestCliVersionArtefact {
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/lib/src/downloads/resource_manager.dart:161]: We can discuss this on Monday as well. Made an attempt but I'm not sure ours serialised helps us in this case.

File: templates/pubspecs/tools/serverpod_cli/pubspec.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (SandPod) [reply templates/pubspecs/tools/serverpod_cli/pubspec.yaml]: Implemented the MockClient without build runner and mockito now so I could remove those two dependencies. Take a look and se what you think.

I also reverted the touch on build runner in the `serverpod_test_shared` project as well since I don't need to interact with build runner in this PR anymore. But I think we might have a problem there supporting Dart SDK 2.19.

File: tools/serverpod_cli/lib/src/util/latest_cli_version.dart [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  latestCliVersion = await pubDevService.tryFetchAndStoreLatestVersion();
+
+  return latestCliVersion;
+}
+
+class LocalStorageService {
+  final String? optionalLocalStoragePath;
+
+  const LocalStorageService({this.optionalLocalStoragePath});
+  LocalStorageService.nonConst({this.optionalLocalStoragePath});
```
Comment (vlidholt) [inline tools/serverpod_cli/lib/src/util/latest_cli_version.dart:33]: Why is this needed?

File: tools/serverpod_cli/lib/src/util/latest_cli_version.dart [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+  var latestCliVersion = await localStorageService.fetchLatestCliVersion();
+
+  if (latestCliVersion != null) return latestCliVersion;
+
+  latestCliVersion = await pubDevService.tryFetchAndStoreLatestVersion();
+
+  return latestCliVersion;
+}
+
+class LocalStorageService {
```
Comment (vlidholt) [inline tools/serverpod_cli/lib/src/util/latest_cli_version.dart:29]: The name feels a bit too generic to be a public class name.

File: tools/serverpod_cli/lib/src/util/latest_cli_version.dart [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+  final String? optionalLocalStoragePath;
+  final http.Client? client;
+  final Duration timeout;
+
+  const PubDevService({
+    this.optionalLocalStoragePath,
+    this.client,
+    this.timeout = LatestCliVersionConstants.pubDevConnectionTimeout,
+  });
+  PubDevService.nonConst({
```
Comment (vlidholt) [inline tools/serverpod_cli/lib/src/util/latest_cli_version.dart:57]: I still don't understand the purpose of the nonConst constructor. 😅

**Review by vlidholt — APPROVED** [review]
Approved with comments fixed.

File: tools/serverpod_cli/lib/src/util/latest_cli_version.dart (hunk 7 above)
Comment (SandPod) [reply tools/serverpod_cli/lib/src/util/latest_cli_version.dart]: It wasn't! Thank you 🙏

**Review by vlidholt — APPROVED** [review]
Approved with minor changes.

File: tools/serverpod_cli/lib/src/downloads/resource_manager.dart [hunk 10]
```diff
@@ -40,22 +43,73 @@ class ResourceManager {
...
+  Future<void> storeLatestCliVersion(
+    CliVersionData cliVersionData, {
+    String? localStoragePath,
+  }) async {
+    localStoragePath ??= localStorageDirectory.path;
+    var latestCliVersionFile = File(p.join(
+        localStoragePath, ResourceManagerConstants.latestVersionFilePath));
+
+    try {
+      if (!latestCliVersionFile.existsSync()) {
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/lib/src/downloads/resource_manager.dart:73]: This also handles creating the folder if it doesn't exist. If we end the process in between we also handle removing corrupted files. So we will keep it.
